hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java

```diff
@@ -112,7 +112,9 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
         if (op.hasAutoRenewAccount()) {
             context.requireKeyOrThrow(op.autoRenewAccountOrThrow(), INVALID_AUTORENEW_ACCOUNT);
         }
-        if (op.hasTreasury()) {
+        // If the transactionID is missing we have contract call,
+        // so we need to verify that the provided treasury is not a zero account.
+        if (op.hasTreasury() && (context.body().hasTransactionID() || !isZeroAccount(op.treasury()))) {
```

Comment:
This test, which also appears down ~L171, is tricky in and of itself (nested opposite boolean operations), and deserves to be pulled out to a method and given a name (and maybe not be a one-liner: maybe the RHS of the `&&` _also_ needs a name).  

Also something similar but different is happening down around L364 - together there's a bit of interesting semantics going on in these rules - and these rules should be pulled out and given names so that we can identify the commonalities and differences and later we might discover other places where one or the other rule is appropriate to use.